rustls/src/server/tls13.rs

```diff
@@ -575,46 +575,35 @@ impl CompleteClientHelloHandling {
             return Err(hs::illegal_param(sess, "client sent duplicate keyshares"));
         }
 
-        let share_groups: Vec<NamedGroup> = shares_ext
-            .iter()
-            .map(|share| share.group)
-            .collect();
-
         let supported_groups = suites::KeyExchange::supported_groups();
-        let chosen_group = supported_groups
+        let chosen_share = shares_ext
             .iter()
-            .find(|group| share_groups.contains(group))
-            .cloned();
-
-        if chosen_group.is_none() {
-            // We don't have a suitable key share.  Choose a suitable group and
-            // send a HelloRetryRequest.
-            let retry_group_maybe = supported_groups
-                .iter()
-                .find(|group| groups_ext.contains(group))
-                .cloned();
-            self.handshake
-                .transcript
-                .add_message(chm);
+            .find(|share| supported_groups.contains(&share.group));
 
-            if let Some(group) = retry_group_maybe {
+        let chosen_share = match chosen_share {
+            Some(share) => share,
+            None => {
                 if self.done_retry {
```

Comment:
I loved the check for `self.done_retry` earlier to avoid wasted work in the error case.

Reply:
Previously this check was conditional on `retry_group_maybe` being `Some`. Is it correct to make it no longer conditional on that check? Does the error message still make sense in that case? Or should this check for `retry_group_maybe.is_some() && self.done_retry`?